prov/util/src/cuda_mem_monitor.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * (C) Copyright 2020 Hewlett Packard Enterprise Development LP
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_mr.h"
+
+#ifdef HAVE_LIBCUDA
+
+#include "ofi_hmem.h"
+
+static int cuda_mm_subscribe(struct ofi_mem_monitor *notifier, const void *addr,
+			     size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	CUresult ret;
+
+	ret = ofi_cuPointerGetAttribute(&hmem_info->cuda_id,
```

Comment:
I had this question in https://github.com/ofiwg/libfabric/issues/5789#issuecomment-608640709 but we did not get to talking about it. If all we need is the buf ID, should we consider extending the device union in fi_mr_attr to include this information? (Thinking about the ~3µs overhead this is going to introduce on each subscribe call).

Reply:
One thing which is not 100% clear is when the buffer ID needs to be allocated. The Nvidia RDMA GPU direct documentation seems to imply that the buffer ID needs to be allocated after device pages are pinned. 

https://docs.nvidia.com/cuda/gpudirect-rdma/index.html#invalidating-based-on-buffer-id